### PR TITLE
rate limiting test cases

### PR DESCRIPTION
**Pull request by psturc**

### CODE REVIEW

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+oc get cm rate-limit-alerts -n redhat-managed-api-operator
+```
+
+2. Verify that level1, level2 and level3 rate-limiting alerts are present
+
+```
+oc get prometheusrules -n redhat-managed-api-marin3r
+```
+
+3. (TODO: modify the alerts configmap - configmap name?)
```
Comment (pmccarthy) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]: Assuming that this test case is performed on a cluster where RHOAM has not been installed via the add-on, we should be able to modify the `sku-limits-managed-api-service` configmap in the RHOAM operator namespace to something like:
```
data:
  rate_limit: |-
    {
      "RHOAM SERVICE SKU": {
        "unit": "minute",
        "requests_per_unit": 20
      }
    }
```
Once the rate limit configs are updated, we should then create a new 3Scale application by logging into 3scale-admin route (login creds can be retrieved from the system-seed secret) and follow the getting started wizard. This provides us with an endpoint that will be rate limited.

Useful command for running X number of requests per second:
```
for i in {1..X}; do curl -i https://api-3scale-apicast-staging.apps.<cluster-id>.devshift.org//?user_key=XXXX; done
```
@sergioifg94 might should have detailed steps on how to trigger each alert

File: test-cases/tests/dashboards/e09-verify-grafana-ratelimiting-dashboard.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+2. RHOAM is installed
+3. SSO IDP configured on a cluster
+
+## Steps
+
+1. Go to redhat-managed-api-customer-monitoring-operator namespace
+2. Go to Routes -> grafana-route and login as customer-admin user using testing-idp
+3. Verify the rate-limiting dashboard is present (TODO: more detailed step)
```
Comment (pmccarthy) [inline test-cases/tests/dashboards/e09-verify-grafana-ratelimiting-dashboard.md]: Perhaps @laurafitzgerald can provide more information on these steps if the Grafana work is added before v0.2.0 is cut

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md (hunk 1 above)
Comment (pmccarthy) [reply test-cases/tests/alerts/c18-validate-api-usage-alerts.md]: It seems that the steps for creating a new 3Scale application are listed in `h22-validate-that-rate-limit-service-is-working-as-expected.md` so we should be able to copy/paste them

Comment (sergioifg94) [reply test-cases/tests/alerts/c18-validate-api-usage-alerts.md]: For now I've only managed to trigger the level 3 one, but I'm gonna try to play with the different configurations to get the other ones

Comment (psturc) [reply test-cases/tests/alerts/c18-validate-api-usage-alerts.md]: Thanks for the additional info guys.

Regarding checking the alerts - maybe this script that resiliency team is using for checking the alerts during performance testing could help us to easily detect the alerts?
https://github.com/integr8ly/integreatly-operator/blob/master/scripts/alerts-during-perf-testing.sh

File: test-cases/tests/dashboards/e09-verify-grafana-ratelimiting-dashboard.md (hunk 2 above)
Comment (laurafitzgerald) [reply test-cases/tests/dashboards/e09-verify-grafana-ratelimiting-dashboard.md]: Just closing this one as this test case has been automated I believe this'll be updated.

Comment (psturc) [reply test-cases/tests/dashboards/e09-verify-grafana-ratelimiting-dashboard.md]: updated here in this commit: https://github.com/psturc/integreatly-operator/pull/2/commits/5f1a0f41e2fd4443ee598499c89bb0f7a52931fa

File: test-cases/tests/dashboards/e09-verify-customer-dashboards-exist-and-shows-a-correct-data.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+## Steps
+
+1. Run the following command and log in as a `customer-admin` user using testing-idp
+
+```
+open "https://$(oc get routes grafana-route -n redhat-rhmi-customer-monitoring-operator -o jsonpath='{.spec.host}')"
+```
+
+2. Verify the "3Scale Api Rate Limiting" dashboard is present and shows some data
```
Comment (laurafitzgerald) [inline test-cases/tests/dashboards/e09-verify-customer-dashboards-exist-and-shows-a-correct-data.md]:
```suggestion
2. Verify the "3Scale API Rate Limiting" dashboard is present and shows some data
```

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      "api-usage-alert-level3": {
+        "ruleName": "Level3ThreeScaleApiUsageThresholdExceeded",
+        "level": "warning",
+        "minRate": "95%",
+        "period": "30m"
+      }
+}
+```
+
+4. Go to `Networking` > `Routes` under `redhat-rhmi-3scale` namespace
```
Comment (pmccarthy) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md:100]: @sergioifg94 would you mind reviewing the steps outlined by @psturc and updating where appropriate please?

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        "level": "warning",
+        "minRate": "90%",
+        "maxRate": "95%",
+        "period": "2h"
+      },
+      "api-usage-alert-level3": {
+        "ruleName": "Level3ThreeScaleApiUsageThresholdExceeded",
+        "level": "warning",
+        "minRate": "95%",
+        "period": "30m"
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]:
```suggestion
        "period": "1m"
```

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        "minRate": "80%",
+        "maxRate": "90%",
+        "period": "1m"
+      },
+      "api-usage-alert-level2": {
+        "ruleName": "Level2ThreeScaleApiUsageThresholdExceeded",
+        "level": "warning",
+        "minRate": "90%",
+        "maxRate": "95%",
+        "period": "2h"
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]:
```suggestion
        "period": "1m"
```

Comment (sergioifg94) [reply test-cases/tests/alerts/c18-validate-api-usage-alerts.md]: I'd set the period for the other levels to 1m beforehand as they're going to be tested as well

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+5. Click on the `zync` route that starts with `https://3scale-admin...`
+6. Go to `Secrets` > `system-seed` under 3Scale namespace and copy the admin password
+7. Go back to 3Scale login page and login
+8. Click on `Ok, how does 3scale work?` and follow the 3Scale wizard to create an API
+9. Once on your API overview page, click on `Integration` on the left, then on `Configuration`
+10. Copy the `example curl for testing` for `Staging-APIcast` and paste into a terminal window
+11. Run the following script to verify that rate limit (20 requests/minute) works correctly:
+
+```
+for i in {1..21}; do <replace-with-example-curl-for-testing>; done
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]:
```suggestion
# Select which number of requests to perform based on the level to trigger
# To trigger level 1 alert
num_req=17
# To trigger level 2 alert
num_req=18
# To trigger level 3 alert
num_req=21
for (( c = 1; c<=$num_req; c++ )); do <replace-with-example-curl-for-testing>; done
```

We need a specific number of requests to trigger level 1 and 2 alerts, as they must be within a range. If we go over the range, only the level 3 alert will fire

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+> Only the last request should fail (TODO: any specific status code?)
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]:
```suggestion
> Only the last request for level 3 should fail with a `428 Too Many Requests` status code
```

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+```
+open "https://$(oc get routes prometheus-route -n redhat-rhmi-middleware-monitoring-operator -o jsonpath='{.spec.host}')"
+```
+
+14. Go to Alerts -> and search for `Level1ThreeScaleApiUsageThresholdExceeded` alert
+15. Click on the alert name -> click on the expression (link) -> Graph
+16. Run the `curl` command against the APIcast endpoint so you reach between 80%-90% rate limit (i.e. amount of curl requests = rate-limit-value \* 0.85)
+17. Verify that the alert graph is showing some data
+18. Repeat the same process for `Level2ThreeScaleApiUsageThresholdExceeded` and `Level3ThreeScaleApiUsageThresholdExceeded` alerts based on their min and max rates and verify that the alerts are firing as expected
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md]:
```suggestion
18. Repeat the same process for `Level2ThreeScaleApiUsageThresholdExceeded` and `Level3ThreeScaleApiUsageThresholdExceeded` alerts based on their min and max rates and verify that the alerts are firing as expected. Make sure to wait at least a minute before running each set of `curl` commands
```

If we run the set of requests too soon, the new requests will be accumulated to the previous one, reaching the rate limit earlier than expected

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+10. Copy the `example curl for testing` for `Staging-APIcast` and paste into a terminal window
+11. Run the following script to verify that rate limit (20 requests/minute) works correctly:
+
+```
+for i in {1..21}; do <replace-with-example-curl-for-testing>; done
+```
+
+> Only the last request should fail (TODO: any specific status code?)
+
+12. TODO: should we update the configmap now to increase the rate limit for test the rate-limit alerts?
```
Comment (sergioifg94) [inline test-cases/tests/alerts/c18-validate-api-usage-alerts.md:122]: I think that we can test the 3 levels without modifying the rate limit configuration. I'm still trying to find the sweet spot to replicate the alerts as consistently as possible, so maybe 20 reqs/minute is not the best, but it's possible to replicate all of the levels with that value

**Review by sergioifg94 — CHANGES_REQUESTED** [review]
Added some comments to the API usage alerts steps

File: test-cases/tests/alerts/c18-validate-api-usage-alerts.md (hunk 10 above)
Comment (pmccarthy) [reply test-cases/tests/alerts/c18-validate-api-usage-alerts.md:122]: We could always look to set the dev limits to 100 requests per minute which would make our calculations much easier

**Comment by psturc** [comment]
@pmccarthy @sergioifg94 thanks for help with writing down the test cases

I'm going to merge this PR now in order to generate RC1 verification tickets in jira and will add a note for myself to update `test-cases/tests/alerts/c18-validate-api-usage-alerts.md` once we have a complete verification steps

**Pull request merged**